tensorflow/compiler/mlir/hlo/tests/lhlo-legalize-to-linalg.mlir

```diff
@@ -1242,3 +1242,53 @@ func @reduce_maximum(%arg: memref<100x10xf32>,
 // CHECK-NEXT: memref.load
 // CHECK-NEXT: linalg.yield
 // CHECK-NEXT: }
+
+// -----
+
+// CHECK-DAG: #[[REDUCE_INPUT_MAP:.*]] = affine_map<(d0, d1) -> (d0, d1)>
+// CHECK-DAG: #[[REDUCE_OUTPUT_MAP:.*]] = affine_map<(d0, d1) -> (d0)>
+// CHECK-LABEL: func @reduce_multiple_operand
+module  {
+  func @reduce_multiple_operand(%arg0: memref<1x8xf32>, %arg1: memref<1x8xi32>,
+                                %arg2: memref<f32>, %arg3: memref<i32>,
+                                %arg4: memref<1xf32>, %arg5: memref<1xi32>) {
+    "lmhlo.reduce"(%arg0, %arg1, %arg2, %arg3, %arg4, %arg5) ( {
+    ^bb0(%arg6: memref<f32>, %arg7: memref<i32>, %arg8: memref<f32>,
+         %arg9: memref<i32>, %arg10: memref<f32>, %arg11: memref<i32>):
+      "lmhlo.add"(%arg6, %arg8, %arg10) : (memref<f32>, memref<f32>, memref<f32>) -> ()
+      "lmhlo.add"(%arg7, %arg9, %arg11) : (memref<i32>, memref<i32>, memref<i32>) -> ()
+      "lmhlo.terminator"() : () -> ()
+    }) {dimensions = dense<1> : tensor<1xi64>} 
+       : (memref<1x8xf32>, memref<1x8xi32>, memref<f32>, memref<i32>, memref<1xf32>, memref<1xi32>) -> ()
+    return
+  }
+}
+// CHECK: %[[INIT_VAL0:.*]] = memref.load %arg2[] : memref<f32>
+// CHECK: linalg.fill(%[[INIT_VAL0]], %arg4) : f32, memref<1xf32> 
+// CHECK: %[[INIT_VAL1:.*]] = memref.load %arg3[] : memref<i32>
+// CHECK: linalg.fill(%[[INIT_VAL1]], %arg5) : i32, memref<1xi32> 
+// CHECK: linalg.generic {
+// CHECK-SAME: indexing_maps = [#[[REDUCE_INPUT_MAP]], #[[REDUCE_INPUT_MAP]], #[[REDUCE_OUTPUT_MAP]], #[[REDUCE_OUTPUT_MAP]]],
+// CHECK-SAME: iterator_types = ["parallel", "reduction"]}
+// CHECK-SAME: ins(%arg0, %arg1 : memref<1x8xf32>, memref<1x8xi32>) outs(%arg4, %arg5 : memref<1xf32>, memref<1xi32>) {
+// CHECK: %2 = memref.alloca() : memref<f32>
```

Comment:
I havent reviewed this yet, but having such allocas here is not what I would expect.

Reply:
I don't think I generated redundant alloca op. The original logic is [here](https://github.com/tensorflow/mlir-hlo/blob/75a20d0492313e5a18eb77d62832ddbe28f0db97/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc#L1590). Do you mean there is no need for temporary variables?

Reply:
I also think allocaop is redundant.I checked the reduceop conversion logic. Allocaop is just for compatibility with the scalarpointwisetostandardconverter. I tried to legalize it's body in the reduce pattern, but this will complicate the problem. I think memref dialect should collapse load and store, but it doesn't. can you give me some suggestions?

Reply:
@d0k is the author of the lowering. Maybe he can comment some on this.

Reply:
It's working as intended. The problem is that we use lmhlo ops inside of the reduce, so if we want to use the lmhlo lowering for that we need to convert from `f32` to `memref<f32>` somehow. That's what the alloca+store is doing. A simple load-store forwarding pass could eliminate all of it after linalg lowering. That's what we did for IREE back then, but I believe IREE is no longer using this lowering and uses mhlo.reduce instead of the lmhlo variant.

This could be avoided if we used mhlo ops inside of lmhlo.reduce, but as far as I know nobody is working on that.

Reply:
Thank you for your answer. I think so. Mhlo and linalg are both SSA forms. From mhlo convert to linalg is simple.